### PR TITLE
Feature/on slide callback

### PR DESCRIPTION
**Pull request by MartijnHarte**
This adds a property that is called when intersection happens so that we can track the user sliding.

It also exposes the first and last fully visible slide indices from the api.

### CODE REVIEW

File: src/Slider.test.tsx [hunk 1]
```diff
@@ -316,15 +309,62 @@ describe('UpsellSlider', () => {
...
+
+            intersectionCallback([
+                { intersectionRatio: 1, target: slides[0] } as unknown as IntersectionObserverEntry,
+                { intersectionRatio: 1, target: slides[1] } as unknown as IntersectionObserverEntry,
+                { intersectionRatio: 1, target: slides[2] } as unknown as IntersectionObserverEntry,
+                { intersectionRatio: .5, target: slides[3] } as unknown as IntersectionObserverEntry,
+                { intersectionRatio: 0, target: slides[4] } as unknown as IntersectionObserverEntry,
+            ], mockIntersectionObserver.mock.instances[0]);
+
+            expect(ref.current!.getFirstFullyVisibleSlideIndex()).toBe(0);
```
Comment (YspMark) [inline src/Slider.test.tsx:366]: You only have one testcase on this function using default state. There is no test on it being updated on slide or button movement.

File: src/Examples/OnSlideExample.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import type { FC, PropsWithChildren, ComponentType } from 'react';
+import { useRef, useState } from 'react';
+import { Slider, SliderTypes } from '../Slider';
+
+const OnSlideExample: FC<typeof Slider extends ComponentType<infer T> ? PropsWithChildren<T> : never> = (sliderProps) => {
```
Comment (iFiz) [inline src/Examples/OnSlideExample.tsx:5]: Nitpicking, but in the current state this still allows `children` to be passed even though they're not rendered

File: Index.tsx [hunk 3]
```diff
@@ -67,6 +68,7 @@ const Index: React.FC = () => {
                 <img draggable={false} loading="lazy" className="demo-slide" width="300" height="200" alt="Slide 5" src="https://via.placeholder.com/300x200/333333"/>
                 <img draggable={false} loading="lazy" className="demo-slide" width="300" height="200" alt="Slide 6" src="https://via.placeholder.com/300x200/333333"/>
             </Slider>
+            <OnSlideExample/>
```
Comment (iFiz) [inline Index.tsx:71]: Nice, we should probably do this for all examples in a separate PR. Things are getting messy real quick with the current setup.

File: src/Examples/OnSlideExample.tsx (hunk 2 above)
Comment (iFiz) [reply src/Examples/OnSlideExample.tsx:5]: Nitpicking part 2: Since it's in the src folder, typescript types will also be generated even though there's not much point in doing so

**Comment by iFiz** [comment]
Please add the new configuration option (`onSlide`), together with the API changes to the README

File: src/Slider.test.tsx [hunk 4]
```diff
@@ -19,20 +23,26 @@ const renderSliderWithDimensions = (clientWidth = 1000, scrollWidth = 2000, scro
 };
 
 describe('UpsellSlider', () => {
+    let mockIntersectionObserver: jest.Mock<IntersectionObserver, ConstructorParameters<typeof IntersectionObserver>>;
     let observeSpy: jest.Mock;
     let disconnectSpy: jest.Mock;
     let scrollToSpy: jest.Mock;
 
+    const getIntersectionObserverInstance = () => mockIntersectionObserver.mock.calls[mockIntersectionObserver.mock.calls.length - 1];
```
[the remark this answers is not in the document]
Comment (MartijnHarte) [reply src/Slider.test.tsx:31]: Yeah sure, every jest mock object has a mock property that stores all the calls being made to that mock. So our mock `mockIntersectionObserver` has all it's calls stored in `mockIntersectionObserver.mock.calls`.

Every time the intersection observer is being instantiated with `new IntersectionObserver` it's constructor call is stored in the mock.calls. This function gives you the latest version of the mock, accessing the latest call. The intersection observer is being disconnected in the cleanup of the useEffect and recreated so we'll want the last one.

This function gives you the constructor parameters of the latest intersection observer instance, allowing you to access the callback function and call it yourself in a test case.

File: src/Slider.test.tsx (hunk 1 above)
Comment (MartijnHarte) [reply src/Slider.test.tsx:366]: That's because they aren't directly related to each other. I thought about it but I think it should be an e2e test, not a unit test.

**Comment by iFiz** [comment]
I removed the target branch, which apparently closed this MR, this was not my intention
